compiler/noirc_frontend/src/hir_def/types.rs

```diff
@@ -1768,13 +1780,49 @@ impl Type {
     ) {
         let mut bindings = TypeBindings::new();
 
-        if let Err(UnificationError) = self.try_unify(expected, &mut bindings) {
-            if !self.try_array_to_slice_coercion(expected, expression, interner) {
-                errors.push(make_error());
-            }
-        } else {
+        if let Ok(()) = self.try_unify(expected, &mut bindings) {
             Type::apply_type_bindings(bindings);
+            return;
+        }
+
+        if self.try_array_to_slice_coercion(expected, expression, interner) {
+            return;
+        }
+
+        // Try to coerce `fn (..) -> T` to `unconstrained fn (..) -> T`
+        if let Some(coerced_self) = self.try_fn_to_unconstrained_fn_coercion(expected) {
+            coerced_self.unify_with_coercions(expected, expression, interner, errors, make_error);
+            return;
+        }
+
+        errors.push(make_error());
+    }
+
+    // If `self` and `expected` are function types, tries to coerce `self` to `expected`.
+    // Returns None if no coercion can be applied, otherwise returns `self` coerced to `expected`.
+    fn try_fn_to_unconstrained_fn_coercion(&self, expected: &Type) -> Option<Type> {
+        // If `self` and `expected` are function types, `self` can be coerced to `expected`
+        // if `self` is unconstrained and `expected` is not. The other way around is an error, though.
+        let (
+            Type::Function(params, ret, env, unconstrained_self),
+            Type::Function(_, _, _, unconstrained_expected),
+        ) = (self.follow_bindings(), expected.follow_bindings())
+        else {
+            return None;
+        };
+
+        // Nothing to coerce
+        if unconstrained_self == unconstrained_expected {
+            return None;
         }
+
+        // unconstrained mismatch: can't coerce
+        if unconstrained_self && !unconstrained_expected {
+            return None;
+        }
+
+        // Cast self unconstrained status to that of expected, so it won't produce an error
+        Some(Type::Function(params, ret, env, unconstrained_expected))
```

Comment:
```suggestion
        (!unconstrained_self && unconstrained_expected).then_some({
            Type::Function(params, ret, env, unconstrained_expected)
        })
```
Simpler to check for only the case we want rather than returning None for each case we don't want